src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                => SyntaxFactory.AnonymousObjectMemberDeclarator(
+                    SyntaxFactory.NameEquals(SyntaxFactory.IdentifierName(name)), SyntaxFactory.IdentifierName(name));
+
+            private static StatementSyntax CreateJoinStatement(JoinClauseSyntax joinClause, StatementSyntax statement)
+            {
+                return SyntaxFactory.Block(
+                                SyntaxFactory.ForEachStatement(
+                                    joinClause.Type ?? VarNameIdentifier,
+                                    joinClause.Identifier,
+                                    joinClause.InExpression,
+                                    SyntaxFactory.Block(
+                                        SyntaxFactory.IfStatement(
+                                            SyntaxFactory.InvocationExpression(
+                                                SyntaxFactory.MemberAccessExpression(
+                                                    SyntaxKind.SimpleMemberAccessExpression,
+                                                    SyntaxFactory.PredefinedType(SyntaxFactory.Token(SyntaxKind.ObjectKeyword)),
+                                                    SyntaxFactory.IdentifierName(nameof(object.Equals))),
+                                                SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(
+                                                    new[] {
+                                                        SyntaxFactory.Argument(joinClause.LeftExpression),
+                                                        SyntaxFactory.Argument(joinClause.RightExpression.WithoutTrailingTrivia())}))),
+                                            statement)))).WithAdditionalAnnotations(Simplifier.Annotation);
+            }
+
+            private class ProcessingState
+            {
+                public ImmutableArray<string> VariableNames { get; private set; }
+
+                // Each clause generates a method converting StatementSyntax into StatementSyntax.
+                public Stack<Func<StatementSyntax, StatementSyntax>> StatementProcessingMethods { get; }
+
+                public Func<ExpressionSyntax, StatementSyntax> LeafExpressionCreationMethod { get; }
+
+                public ProcessingState(Func<ExpressionSyntax, StatementSyntax> leafExpressionCreationMethod)
+                {
+                    VariableNames = ImmutableArray<string>.Empty;
+                    StatementProcessingMethods = new Stack<Func<StatementSyntax, StatementSyntax>>();
+                    LeafExpressionCreationMethod = leafExpressionCreationMethod;
+                }
+
+                public void AddVariableName(string variableName)
+                    => VariableNames = VariableNames.Concat(variableName);
+
+                public void AddStatementProcessingMethod(Func<StatementSyntax, StatementSyntax> method)
```

Comment:
>AddStatementProcessingMethod [](start = 28, length = 28)

It feels that the the stack of delegates doesn't really buy's us much and actually makes code more complicated and harder to reason about (debugging probably going to be a major pain). I think all these transformations can be done directly when we simply unwrap the stack of query clauses to translate.